core/cloudflow-spark-testkit/src/main/scala/cloudflow/spark/testkit/SparkStreamletTestkit.scala

```diff
@@ -256,3 +257,55 @@ case class SparkOutletTap[T: Encoder](
   // get results from memory sink
   def asCollection(session: SparkSession): Seq[T] = session.sql(s"select * from $queryName").as[T].collect()
 }
+case class ExecutionReport(totalRows: Long, totalQueries: Int, failures: Seq[String]) {
+  def hasFailures = failures.nonEmpty
+}
+
+class QueryExecutionMonitor()(implicit ec: ExecutionContext) extends StreamingQueryListener {
+  @volatile var status: Map[UUID, QueryState] = Map()
+  @volatile var dataRows: Map[UUID, Long]     = Map()
+
+  sealed trait QueryState
+  case object Started extends QueryState
+  case class Terminated(exception: Option[String]) extends QueryState {
+    def isFailed = exception.nonEmpty
+  }
+
+  def executionReport: ExecutionReport = {
+    val exceptions = status.values.collect { case Terminated(Some(reason)) => reason }
+    ExecutionReport(dataRows.values.sum, dataRows.keys.size, exceptions.toSeq)
+  }
+
+  def waitForData(): Future[Long] = {
+    def hasData   = dataRows.nonEmpty && dataRows.forall { case (_, rows) => rows > 0 }
+    def totalRows = dataRows.values.sum
+    Future {
+      while (!hasData) {
+        Thread.sleep(1.second.toMillis)
```

Comment:
Better to use promise and future here ( create a promise, complete it when data arrives, await on future of promise here, no need to sleep)

Reply:
Good point! Thanks.

Reply:
I created a `poll` for our prod code, but I find it's overkill here.
See also: https://github.com/lightbend/cloudflow/pull/222/files#diff-80785b7e8093f0a01524d8c1768e4347R126

Reply:
I don't understand really cause the hasData will flip once, so you can use promise future and no poll? Am I missing something?

Reply:
@RayRoestenburg one option I could think of is to move this check to `onQueryProgress` but that check code looks intrusive there as it mixes the work of the progress callback with the data availability reporting. 
It might be less obvious to maintain later.
Would you have a suggestion?

Reply:
You can create the promise in the body of the QueryExecutionMonitor, pass the future of it here, complete promise on first data (probably in onQueryProgress)? I don't see the problem with that (yet)